src/benchmarks/micro/corefx/System.Threading.Tasks.Dataflow/Perf.Dataflow.cs

```diff
@@ -0,0 +1,384 @@
+            await Task.WhenAll(
+                Post(block.Target1),
+                Post(block.Target2),
+                Receive()
+            );
+        }
+
+        [Benchmark(OperationsPerInvoke = 100_000)]
+        public async Task SendAsyncTwiceReceiveAsyncOnceParallel()
+        {
+            await Task.WhenAll(
+                SendAsync(block.Target1),
+                SendAsync(block.Target2),
+                ReceiveAsync()
+            );
+        }
+    }
+    #endregion
+
+    #region  Abstract classes
+    [BenchmarkCategory(Categories.CoreFX)]
+    public abstract class PerfTests<T> where T : IDataflowBlock
+    {
+        protected T block;
+
+        public abstract T CreateBlock();
+
+        [GlobalSetup]
+        public void BlockSetup()
+        {
+            block = CreateBlock();
+        }
+
+        [Benchmark]
+        public async Task Completion()
+        {
+            block.Complete();
+            await block.Completion;
+        }
+
+        protected static Task Post(ITargetBlock<int> target, bool retry = false) => Task.Run(() =>
+        {
+            for (int i = 0; i < 100_000; i++)
+            {
+                while (!target.Post(i) && retry) ;
+            }
+        });
+
+        protected static Task Receive<U>(ISourceBlock<U> source, int receiveSize = 1) => Task.Run(() =>
+        {
+            for (int i = 0; i < 100_000 / receiveSize; i++)
+            {
+                source.Receive();
+            }
+        });
+
+        protected static async Task SendAsync(ITargetBlock<int> target)
+        {
+            for (int i = 0; i < 100_000; i++)
+            {
+                await target.SendAsync(i);
+            }
+        }
+
+        protected static async Task ReceiveAsync<U>(ISourceBlock<U> source, int receiveSize = 1)
+        {
+            for (int i = 0; i < 100_000 / receiveSize; i++)
+            {
+                await source.ReceiveAsync();
+            }
+        }
+    }
+
+    [BenchmarkCategory(Categories.CoreFX)]
+    public abstract class SourceBlockPerfTests<T, U> : PerfTests<T> where T : ISourceBlock<U>
+    {
+        protected virtual int ReceiveSize { get; } = 1;
+        protected Task Receive() => Receive(block, ReceiveSize);
+        protected Task ReceiveAsync() => ReceiveAsync(block, ReceiveSize);
+    }
+
+    [BenchmarkCategory(Categories.CoreFX)]
+    public abstract class DefaultTargetPerfTests : TargetPerfTests<ITargetBlock<int>> { }
+
+    [BenchmarkCategory(Categories.CoreFX)]
+    public abstract class TargetPerfTests<T> : PerfTests<T> where T : ITargetBlock<int>
+    {
+        [Benchmark(OperationsPerInvoke = 100_000)]
+        public Task Post() => Post(block);
+
+        [Benchmark(OperationsPerInvoke = 100_000)]
+        public Task SendAsync() => SendAsync(block);
+    }
+
+    [BenchmarkCategory(Categories.CoreFX)]
+    public abstract class DefaultBoundedPropagatorPerfTests : BoundedPropagatorPerfTests<IPropagatorBlock<int, int>, int> { }
```

Comment:
could you please remove this class to reduce the class hierarchy depth and use `BoundedPropagatorPerfTests<IPropagatorBlock<int, int>, int>` instead of `DefaultBoundedPropagatorPerfTests` when defining inheritor classes?